service/storage_service.cc

```diff
@@ -4721,11 +4722,15 @@ future<> storage_service::do_cluster_cleanup() {
 
 future<sstring> storage_service::wait_for_topology_request_completion(utils::UUID id, bool require_entry) {
     while (true) {
+        auto c = _topology_state_machine.reload_count;
         auto [done, error] = co_await  _sys_ks.local().get_topology_request_state(id, require_entry);
         if (done) {
             co_return error;
         }
-        co_await _topology_state_machine.event.when();
+        if (c == _topology_state_machine.reload_count) {
+            // wait only if the state was not reloaded while we were preempted
+            co_await _topology_state_machine.event.when();
```

Comment:
Might be good to also put a (debug?) log line just before (and maybe even after) or at least somewhere in the loop (and when exiting it) to know when we get stuck here.

But could as well be a follow-up commit (together with enabling the log in the tests)